Update README.md (#2)
rancher-sandbox/rancher-desktop-host-resolver: README.md

# Host Resolver

> A stub DNS resolver that runs on the host machine on Linux, macOS, and Windows. The main goal behind this stub resolver is more robuset handling of domain name resolutions when VPN split tunnle is setup.  

Reviewer: ```suggestion
A stub DNS resolver that runs on the host machine on Linux, macOS, and Windows. The main goal behind this stub resolver is more robuset handling of domain name resolutions when VPN split tunnel is setup.  
```

> ![b822615f-af15-4997-981b-53a6f1153d81 sketchpad (1)](https://user-images.githubusercontent.com/10409174/161135610-6418fe36-cde0-46d2-8e88-11cab9a3b3a2.svg)

Reviewer: The URL shows a page with 2 black boxes to me. Does it work when rendered as part of Markdown?

## Run

```bash
/host-resolver run -a 127.0.0.1 -t 54 -u 53 -c "host.rd.internal=111.111.111.111,host2.rd.internal=222.222.222.222"
```
NOTE: If ports are not provided, host resolver will listen on random ports.

## Test

```bash
docker build -t host-resolver:latest . && docker run --dns 127.0.0.1 -it host-resolver:latest
```

> Note: Run with `--dns` flag, this overrides the `/etc/resolv.conf` in the running container. 

Reviewer: This doesn't include any test command.

The "test" also only works for static names defined via the `-c` option. Any other names will result in a recursive lookup until you hit a timeout.

# Host Resolver

> A stub DNS resolver that runs on the host machine on Linux, macOS, and Windows. The main goal behind this stub resolver is more robuset handling of domain name resolutions when VPN split tunnel is setup.

Reviewer 2: ```suggestion
A stub DNS resolver that runs on the host machine on Linux, macOS, and Windows. The main goal behind this stub resolver is more robust handling of domain name resolutions when using a VPN split tunnel setup.
```

## How Does It Work In Rancher-Desktop?

> Below is the current architecture of Windows version of Rancher Desktop using host-resolver:

Reviewer 2: ```suggestion
Below is the current architecture of Rancher Desktop when using host-resolver on Windows:
```

```mermaid
flowchart  LR;
 subgraph Host

>  isp("Public DNS ISP")

Reviewer 2: ```suggestion
 isp("Public DNS")
```
Thinks like `8.8.8.8` are not ISP, but people use them (e.g. because Chrome).

 corp("Corporate VPN")
 resolver{"Host Resolver (Host Process)"}
 api(("win32 API"))
 resolver <-.->  api
 api  <--->  isp
 api  <--->  corp
 end

>  subgraph VM

Reviewer 2: ```suggestion
 subgraph VM("WSL VM")
```

 peer{"Host Resolver \n(Peer Process)"}
 c1("container 1")
 c2("container 2")
 c1 <----> |tcp/udp :53| peer
 c2 <----> |tcp/udp :53| peer
 end
 peer  <---> |AF_VSOCK| resolver
```

> ## Run

Reviewer 2: ```suggestion
## Running host-resolver
```

> You can run host-resolver in a few different mode:

Reviewer 2: ```suggestion
You can run host-resolver in a few different modes:
```

### 1) DNS Stub Resolver Over AF_VSOCK

In WSL Distro:
```
/host-resolver vsock-peer
```
In Windows Host:
```

> /host-resolver vscok-host --built-in-hosts host.rancher-desktop.internal=111.111.111.111

Reviewer 2: ```suggestion
/host-resolver vsock-host --built-in-hosts host.rancher-desktop.internal=111.111.111.111
```
Using `host.domain.example` might be better (that's a TLD reserved for documentation; see [RFC 6761](https://datatracker.ietf.org/doc/html/rfc6761#section-6.5).). Similarly, use `192.0.2.x` per [RFC 5737](https://datatracker.ietf.org/doc/html/rfc5737#section-3).

```
### 2) Standalone Server
```bash

> /host-resolver run -a 127.0.0.1 -t 54 -u 53 -c "host.rd.internal=111.111.111.111,host2.rd.internal=222.222.222.222"

Reviewer 2: We should use long options in scripts & documentation where available. (For example, `--listen-address` instead of `-a`.)

Reviewer 2: The subcommand was renamed to `standalone`.

```
NOTE: If ports are not provided, host resolver will listen on random ports.

## Test

You can run the tests in the container by running:
```bash
docker build -t host-resolver:latest . && docker run --dns 127.0.0.1 -it host-resolver:latest
```

> Note: Run with `--dns` flag, this overrides the `/etc/resolv.conf` in the running container.

Reviewer 2: ```suggestion
**Note:** Run with `--dns` flag, this overrides the `/etc/resolv.conf` in the running container.
```
Consider rephrasing too… I'm thinking something like _the `--dns` flag is required to override the DNS resolver used in the container_.

# Host Resolver
A stub DNS resolver that runs on the host machine on Linux, macOS, and Windows. The main goal behind this stub resolver is more robuset handling of domain name resolutions when VPN split tunnel is setup.  

> ![b822615f-af15-4997-981b-53a6f1153d81 sketchpad (1)](https://user-images.githubusercontent.com/10409174/161309152-e048d73f-fbe6-42a2-a409-c29a7de7f03a.jpeg)

Author: mermaid UI generator online is broken. Didn't want to waste too much time figuring out what is going on.

Reviewer 2: ```suggestion
```mermaid
flowchart LR
  subgraph Host
  isp["Public DNS ISP"]
  corp["Corporate VPN"]
  resolver["Host Resolver"]
  isp --- resolver
  corp --- resolver
  end
  subgraph VM
  dns["DNS"]
  container1["container 1"]
  container2["container 2"]
  dns --- container1
  dns --- container2
  end
  resolver --- dns
```
Sorry you'll need to add your own close quote block, GitHub doesn't know how to deal with nested triple-backticks correctly.

Author: Thanks @mook-as!
